infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -452,13 +467,43 @@ public String evaluateLegacyNamespaceName(NamespaceResolutionContext resolutionC
    */
   public String evaluateNamespaceName(NamespaceResolutionContext resolutionCtx)
       throws InfrastructureException {
-    String namespace = evalPlaceholders(defaultNamespaceName, resolutionCtx);
+    String namespace;
+    Optional<Pair<String, String>> namespaceOptional =
+        allowUserDefinedNamespaces ? getPreferencesNamespaceName(resolutionCtx) : Optional.empty();
+    if (namespaceOptional.isEmpty() || !isStoredTemplateValid(namespaceOptional.get().second)) {
+      namespace = evalPlaceholders(defaultNamespaceName, resolutionCtx);
+    } else {
+      namespace = namespaceOptional.get().first;
+    }
+
+    if (!NamespaceNameValidator.isValid(namespace)) {
+      Optional<KubernetesNamespaceMeta> namespaceMetaOptional;
+      String normalizedNamespace = normalizeNamespaceName(namespace);
+      if (normalizedNamespace.isEmpty()) {
+        throw new InfrastructureException(
+            format(
+                "Evaluated empty namespace name for workspace %s", resolutionCtx.getWorkspaceId()));
+      }
+      do {
+        normalizedNamespace =
+            normalizedNamespace
+                .substring(0, Math.min(55, normalizedNamespace.length()))
+                .concat(NameGenerator.generate("-", 6));
+        namespaceMetaOptional = fetchNamespace(normalizedNamespace);
+      } while (namespaceMetaOptional.isPresent());
+      namespace = normalizedNamespace;
+    }
 
     LOG.debug(
         "Evaluated the namespace for workspace {} using the namespace default to {}",
         resolutionCtx.getWorkspaceId(),
         namespace);
 
+    if (allowUserDefinedNamespaces
```

Comment:
IMHO storing the value in the user prefs would warrant a separate method. Are you sure this method is always called from the contexts that have a user id recorded in their environment context?

Reply:
I have changed getting user from NamespaceResolutionContext instead of EnvCtx. Is thet waht you meant ?